src/main/docs/guide/config/evaluatedExpressions.adoc

```diff
@@ -0,0 +1,376 @@
+constructs, it is in general not recommended to place complex logic inside it as there are usually better ways to
+achieve the same result.
+
+== Evaluated Expression language reference
+
+The Evaluated Expressions syntax supports the following functionality:
+
+* Literal Values
+* Math Operators
+* Comparison Operators
+* Logical Operators
+* Ternary Operator
+* Type References
+* Method Invocation
+* Property Access
+
+=== Literal Values
+
+The following types of literal values are supported:
+
+* `null`
+* boolean values (`true`, `false`)
+* strings, which need to be surrounded with single quotation mark (`'`)
+* numeric values (`int`, `long`, `float`, `double`)
+
+Integer and Long values can also be specified in hexadecimal or octal notation. Float and Double values can also be
+specified in exponential notation. All numeric values can be negative as well.
+
+.Literal values examples
+[source]
+----
+#{ null }
+#{ true }
+#{ 'string value' }
+#{ 10 }
+#{ 0xFFL }
+#{ 10L }
+#{ .123f }
+#{ 1E+1d }
+#{ 123D }
+----
+
+=== Math Operators
+
+The supported mathematical operators are `+`, `-`, `*`, `/`, `%`, `^`. Math operators can only be applied to numeric
+values (except `+` which can be used for string concatenation as well). Mathematical operations are performed in order
+enforced by standard operator precedence. You can also change evaluation order by using brackets `()`.
+
+`/` and `%` operators can be aliased by `div` and `mod` keywords respectively.
+
+.Math operators examples
+[source]
+----
+#{ 1 + 2 }             // 3
+#{ 'a' + 'b' + 'c' }   // 'abc'
+#{ 7 - 3 }             // 4
+#{ 7 * 3 }             // 21
+#{ 7 * ( 3 + 1) }      // 28
+
+#{ 15 / 3 }            // 5
+#{ 15 div 3 }          // 5
+
+#{ 15 % 3 }            // 0
+#{ 15 mod 3 }          // 0
+
+// Unlike in Java, ^ operator means exponentiation
+#{ 3 ^ 2 }             // 9
+----
+
+=== Comparison Operators
+
+The following comparison operators are supported: `==`, `!=`, `>`, `<`, `>=`, `\<=`, `matches`
+Comparison operations are performed in order enforced by standard operator precedence.
+You can also change evaluation order by using brackets `()`.
+
+Equality check is supported for both primitive types and objects. It is performed using `Object.equals()` method.
+
+`>`, `<`, `>=`, `\<=` operations can only be applied to numeric types.
+
+`matches` keyword can be used to determine whether a string matches provided regular expression which has to
+be specified as string literal. The regular expression itself will be checked for validity at compilation time.
+
+.Comparison operators examples
+[source]
+----
+#{ 1 + 2 == 3 }         // true
+#{ 'abc' != 'abc' }     // false
+#{ 7 > 3 }              // true
+#{ 7 < 3 }              // false
+#{ 7 >= 7 }             // true
+#{ 7 <= 8 }             // false
+
+#{ 'AbC' matches '[A-Za-z*'  }      // Compilation failure
+#{ 'AbC' matches '[A-Za-z]*'  }     // true
+#{ 'AbC' matches '[a-z]*'  }        // false
```

Comment:
what happens with regex escaping? Can we support `/.+/` syntax?

Reply:
regex escaping works in the same as as for `@Requires(property = ..., pattern = ...)`. So in annotation it looks like this
```
"#{ '123' matches '\\d+' }"
```

`/.+/` syntax can be supported, will add it to the list in #8323